specs/merkle-tree.md

```diff
@@ -0,0 +1,142 @@
+
+To resolve these types of problems we outline a set of Merkle tree invariants below that define what a 'valid' Merkle tree looks like. We expect relayers to rely on Merkle trees following the invariants set out here. We assume that relayers will refuse to execute Merkle trees that do not follow these invariants.
+
+## High-Level Merkle Tree Invariants
+
+### 1. Must be executable on all chains for which there is at least one Merkle leaf.
+Assumption: The input Merkle leaf(s) data satisfies the on-chain conditions in the corresponding SphinxModuleProxy contract(s). See the [SphinxModuleProxy specification](https://github.com/sphinx-labs/sphinx/blob/feature/pre-audit/specs/sphinx-module-proxy.md#function-level-invariants) for more information.
+
+### 2. Each leaf on a chain must have an index field with a value that sequentially increments by 1
+To ensure that `EXECUTE` leaves are executed in the correct order, all Merkle tree leaves must have an `index` field that starts at 0 on each chain and sequentially increments by 1.
+
+### 3. Every leaf `chainId` and `index` combination must be unique within the tree
+Constructing a tree with multiple leaves with the same chain ID and index would make it ambiguous which leaf should be executed, and would result in a tree that cannot be fully executed. So we impose the restriction that in a valid Merkle tree there must be at most one leaf with each chain ID and index combination which ensures there is unambiguous ordering of leaves within the tree.
+
+### 4. There must be exactly one `APPROVE` leaf or `CANCEL` leaf per chain
+Constructing a tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain would lead to an ambiguous situation for the relayer where it is unclear which deployment should be approved or canceled. So we impose the restriction that in a valid Merkle tree there must be exactly one `APPROVE` or `CANCEL` leaf per chain.
+
+If the `arbitraryChain` field is set to true, then we consider this approval leaf to apply to all chains. Therefore if `arbitraryChain` = true, there should be exactly one `APPROVE` leaf in the entire tree and no `CANCEL` leaves.
+
+### 5. Merkle tree leaves must be ordered in the tree by the leaf's `index` and `chainId` fields ascending [^2]
+Sphinx Merkle tree leaves use an explicit `index` field in the leaf rather than an implicit index determined by the leaves position in the tree. This is because the Sphinx Merkle tree is intended to be executed across multiple networks and not all leaves will be executed on all networks. We impose the restriction the tree leaves will be ordered by their `index` and `chainId` fields ascending.
+
+So for example, if we intended to run a single transaction on Goerli (5), OP Goerli (420), and Polygon Mumbai (80001). Then the leaves are are expected to be ordered like so:
```

Comment:
What do you mean by a "single transaction"? Do you mean a single `EXECUTE` leaf?